### ISSUE TITLE
Hot reload (in VS Code) is broken for Serverpod server

### ISSUE THREAD

**Issue by lukehutch** [serverpod/serverpod#1748]
I think (but I'm not sure) that a few months ago, hot reload used to work in VS Code when running the Serverpod server in the debugger. It no longer works. When I make changes in a running local server, and hit save, the changes don't go live on the server.

I don't even know how to begin debugging this, so suggestions are welcome!

**Comment by yahu1031** [comment]
So basically, Dart doesn't support hot reloads, whereas Flutter does. So need to restart(hot) the server every time. I hope this feature will be added to the extension soon. I can give it a shot.

**Comment by SandPod** [comment]
Not really sure we actually ever supported it natively.

But there is a class here, [link to source](https://github.com/serverpod/serverpod/blob/0b47da20ebb6ceb47bfa0eb8b6f8089d120df222/packages/serverpod/lib/src/hot_reload/hot_reload.dart#L9), that suggest that you can hot reload the server manually. It is hooked up to an insights endpoint to support hot reloading through and endpoint call. But I don't think we have any tests for it so I won't vouch for how well it functions.

**Comment by lukehutch** [comment]
Right, I noticed that endpoint. But I'm more interested in IDE support.

@yahu1031 why does Dart not have this, if Flutter does?

**Comment by yahu1031** [comment]
@lukehutch ahhhh! Not sure about why, but there is a flag`--observe:<PORT>` with combination of `--no_dds`(optional) and a package that help with it to observe the changes in the server code. This is somewhere I have read. I will reconfirm it with checking with some code.

**Comment by yahu1031** [comment]
#### Update ####
And yes as of the link provided, once the server starts watching for changes for `x secs` of debounce interval, it should call the source reload as [mentioned here](https://github.com/serverpod/serverpod/blob/0b47da20ebb6ceb47bfa0eb8b6f8089d120df222/packages/serverpod/lib/src/hot_reload/hot_reload.dart#L35-L38). This may come when `serverpod run` [comes back to life](https://pub.dev/packages/serverpod_cli/changelog#:~:text=Removes%20the%20serverpod%20run%20command%20in%20favor%20for%20continuous%20generation.)(Hopping so).

@lukehutch Here is some wiki about [Dart Hot-Reload](https://github.com/dart-lang/sdk/wiki/Hot-reload).

**Comment by vlidholt** [comment]
Yeah, we haven't done the hot reload/restart yet. But it would surely be a nice addition!

**Comment by Katekko** [comment]
Just commenting to show that are more people interested in this feature. btw good work!

**Comment by yahu1031** [comment]
But this actually needs a hook to implement the reload functionality. I think it is like hot-restart - ~hot-reload~. The dart reload needs an external library/package to do it as mentioned above. I was working on it, and I have held this to explore more options.

**Comment by dkbast** [comment]
Have you heard of "fire" - this looks promising, its a different approach, but could be viable for integrating hot restart directly into the serverpod cli without the need for an IDE hook: https://github.com/ykmnkmi/fire.dart

**Comment by Kiruel** [comment]
In the meantime you can still use the following:

Install nodemon via:
`npm install -g nodemon`
Then:
`nodemon -x "dart run bin/main.dart --apply-migrations" -e dart`

**Comment by dinko7** [comment]
Gazzele [implementation](https://github.com/intales/gazelle/blob/082366e9f12e69d9204984b77fdbd4c28117dae8/packages/gazelle_cli/lib/commands/run/run_project.dart#L68) of hot reload relies on [Hotreloader](https://pub.dev/packages/hotreloader) package, similar to "fire" @dkbast suggested.

**Comment by yahu1031** [comment]
yes, I have seen that already. Would be a good idea, but we already have inbuild reload functionality in serverpod but it was disabled now 😜 @dinko7

**Comment by omoney1** [comment]
> yes, I have seen that already. Would be a good idea, but we already have inbuild reload functionality in serverpod but it was disabled now 😜 [@dinko7](https://github.com/dinko7)

what do you mean with 'we already have inbuild reload functionality in serverpod but it was disabled now'? Serverpod used to have hot reload?

**Comment by yahu1031** [comment]
> Serverpod used to have hot reload? If yes, why does not have it any more? I'm building a Web site with the dart jaspr framework and I'm using Serverpod as the backend for it, however I found it very frustrating to develop when I don't have hot reload.

Here is the code for [hot-reloading](https://github.com/serverpod/serverpod/blob/0b47da20ebb6ceb47bfa0eb8b6f8089d120df222/packages/serverpod/lib/src/hot_reload/hot_reload.dart#L34-L57). Not sure why it was deprecated. @vlidholt has to answer why it was done. I think I have seen the reason why it was. Let me check and update it.
